packages/blade/src/components/Tabs/utils.ts

```diff
@@ -0,0 +1,85 @@
+import React from 'react';
+import type { TabsProps } from './types';
+import { getComponentId } from '~utils/isValidAllowedChildren';
+import { logger, throwBladeError } from '~utils/logger';
+import { useIsomorphicLayoutEffect } from '~utils/useIsomorphicLayoutEffect';
+
+const iconSizeMap = {
+  medium: 'medium',
+  large: 'large',
+} as const;
+
+const badgeSizeMap = {
+  medium: 'small',
+  large: 'medium',
+} as const;
+
+const counterSizeMap = {
+  medium: 'small',
+  large: 'small',
+} as const;
+
+const propRestrictionMap = {
+  Badge: {
+    medium: {
+      size: badgeSizeMap.medium,
+    },
+    large: {
+      size: badgeSizeMap.large,
+    },
+  },
+  Counter: {
+    medium: {
+      size: counterSizeMap.medium,
+    },
+    large: {
+      size: counterSizeMap.large,
+    },
+  },
+} as const;
+
+type TrailingComponents = keyof typeof propRestrictionMap;
+
+const useTabsItemPropRestriction = (
+  trailing: React.ReactNode,
+  tabItemSize: NonNullable<TabsProps['size']>,
+): React.ReactNode => {
+  const [
+    validatedTrailingComponent,
+    setValidatedTrailingComponent,
+  ] = React.useState<React.ReactElement | null>(null);
+
+  // validate and restrict sub component props in trailing prop
+  useIsomorphicLayoutEffect(() => {
```

Comment:
do we use this kind of method for all the restrictions? haven't seen in the previous components 🤔

Reply:
BottomSheet/BaseHeader uses this. 

It's same as any other inline runtime checks that we do, just extracted to a hook because this is a lot of code.